File: udemy/ds-and-a/csharp-solutions/DSAndA/LinkedLists/LinkedList.cs
using System;
using System.Collections.Generic;

namespace DSAndA.LinkedLists
{
    public class LinkedList<T> where T : IComparable<T>
    {
        private Node<T> head;

        public int Lenght { get; private set; }

        public override string ToString()
        {
            string nodes = string.Empty;

            Node<T> current = this.head;

            while (current != null)
            {
                nodes += $"{current.ToString()},";

                current = current.NextNode;
            }

            return nodes;
        }

        public void InsertAtHead(T data)
        {
            Node<T> nodeToBeInserted = new Node<T>(data) { NextNode = this.head };

            this.head = nodeToBeInserted;

            ++this.Lenght;
        }

        public void DeleteFromHead()
        {
            this.head = this.head.NextNode;

            --this.Lenght;
        }

        /// <summary>
        /// Algorithm to delete a node from somewhere in the middle of a SINGLY linked list, given that you have access only to this node (and of course the subsequent nodes).
        /// </summary>
        /// <param name="node"></param>
        public void DeleteFromMiddle(Node<T> node)
        {
            node.OverwriteData(node.NextNode.GetData());

            node.NextNode = node.NextNode.NextNode;
        }

        public Node<T> Search(T data)
        {
            Node<T> current = this.head;

            while (current != null)
            {
                if (current.GetData().CompareTo(data) == 0)
                    return current;

                current = current.NextNode;
            }

            return null;
        }

        public void AppendLastNodesToBeginning(int n)
        {
            Node<T> current = this.head;
            int currentPosition = 1;

            // Example: 1 -> 2 -> 3
            // Iterate through the list until reaching the (count - n) element = 2
            while (currentPosition++ != this.Lenght - n)
                current = current.NextNode;

            // Save a reference to the current element as toBeLast
            Node<T> toBeLast = current;

            // Go one more element forward
            current = current.NextNode;

            // Save a reference to the current element as toBeHead
            Node<T> toBeHead = current;

            // Go to the last element of the list
            while (current.NextNode != null)
                current = current.NextNode;

            // Connect last element to head: 3 -> 1 -> 2 -> null
            current.NextNode = this.head;

            // Make toBeHead the new head
            this.head = toBeHead;

            // Connect toBeLast.Next to null: 1 -> 2 -> null
            toBeLast.NextNode = null;
        }

        public void Reverse()
        {
            LinkedList<T> reversedLinkedList = new LinkedList<T>();

            Node<T> current = this.head;

            while (current != null)
            {
                reversedLinkedList.InsertAtHead(current.GetData());

                current = current.NextNode;
            }

            this.head = reversedLinkedList.head;
        }

        /// <summary>
        /// With an input of k=1, the last element is returned; with k=2, the second last element; and on so on.
        /// 
        /// Example: 1 -> 2 -> 3 -> 4
        /// 1 = 4th last element => k = 4 (Lenght + 1 - Position = 4 + 1 - 1 = 4)
        /// 4 = 1st last element => k = 1 (Lenght + 1 - Position = 4 + 1 - 4 = 1)
        /// </summary>
        /// <param name="k"></param>
        /// <returns></returns>
        public Node<T> FindKth(int k)
        {
            Node<T> current = this.head;

            int kth = this.Lenght;

            while (kth-- != k && current != null)
                current = current.NextNode;

            return current;
        }

        /// <summary>
        /// One single while loop to traverse the list, using a dictionary to detect the duplicates and remove them.
        /// So, the time complexity is O(n).
        /// </summary>
        public void RemoveDuplicates()
        {
            Node<T> current = this.head;

            Dictionary<T, bool> presenceDictionary = new Dictionary<T, bool>
            {
                [current.GetData()] = true
            };

            Node<T> previous = current;
            current = current.NextNode;

            while (current != null)
            {
                if (presenceDictionary.ContainsKey(current.GetData()))
                    previous.NextNode = current.NextNode; // Removing the duplicate...
                else
                {
                    previous = current;
                    presenceDictionary[current.GetData()] = true;
                }

                current = current.NextNode;
            }
        }

        /// <summary>
        /// One main while loop to traverse the list, and another nested while loop to remove the duplicates for the current element (if any).
        /// So, the time complexity is O(n squared).
        /// </summary>
        public void RemoveDuplicates_NoAdditionalMemoryApproach()
        {
            Node<T> current = this.head;

            while (current != null)
            {
                Node<T> duplicatesLoopPrevious = current;
                Node<T> duplicatesLoopCurrent = current.NextNode;

                while (duplicatesLoopCurrent != null)
                {
                    if (duplicatesLoopCurrent.GetData().CompareTo(current.GetData()) == 0)
                        duplicatesLoopPrevious.NextNode = duplicatesLoopCurrent.NextNode; // Removing the duplicate...
                    else
                        duplicatesLoopPrevious = duplicatesLoopCurrent;

                    duplicatesLoopCurrent = duplicatesLoopCurrent.NextNode;
                }

                current = current.NextNode;
            }
        }

        public bool IsCircular()
        {
            Node<T> current = this.head;

            while (current != null)
            {
                if (current.NextNode != null && current.NextNode == this.head)
                    return true;

                current = current.NextNode;
            }

            return false;
        }
    }
}

File: udemy/ds-and-a/javascript-solutions/src/LinkedLists/LinkedList.js
const Node = require("./Node");

class LinkedList {
    constructor() {
        this.head = null;
    }

    // O(1)
    insertAtHead(data) {
        let newNode = new Node(data);
        newNode.nextNode = this.head;
        this.head = newNode;

        return newNode;
    }

    // O(n)
    insertInSorted(data) {
        if (!this.head) {
            return this.insertAtHead(data);
        }

        let newNode = new Node(data);

        if (data <= this.head.data) {
            newNode.nextNode = this.head;
            this.head = newNode;

            return;
        }

        let current = this.head;

        while (current.nextNode != null) {
            if (data <= current.nextNode.data) {
                newNode.nextNode = current.nextNode;
                current.nextNode = newNode;

                return;
            }

            current = current.nextNode;
        }

        return;
    }

    // O(1)
    deleteFromHead() {
        this.head = this.head.nextNode;
    }

    length() {
        let result = 0;
        let current = this.head;

        while (current != null) {
            result++;

            current = current.nextNode;
        }

        return result;
    }

    // O(n)
    find(data) {
        let current = this.head;

        while (current != null) {
            if (current.data === data) {
                return current;
            }

            current = current.nextNode;
        }

        return null;
    }

    appendLastNodesToBeginning(n) {
        const length = this.length();
        let prevNodeOfLastNodes = null;
        let current = this.head;

        // Sample to use as a visual reference:
        // 0 (head) -> 1 -> 2 -> 3 -> 4 -> null

        // Position current pointer at "length - n - 1" element...
        for (let i = 0; i < length - n - 1; i++) {
            current = current.nextNode;
        }

        // Save "length - n - 1" position...
        prevNodeOfLastNodes = current;

        // Position current pointer at last element...
        while (current.nextNode != null) {
            current = current.nextNode;
        }

        // Set last.next to head...
        current.nextNode = this.head;

        // Set head to "length - n - 1".next...
        this.head = prevNodeOfLastNodes.nextNode;

        // Set "length - n - 1".next to null...
        prevNodeOfLastNodes.nextNode = null;
    }

    reverse() {
        let linkedList = new LinkedList();

        let current = this.head;

        while (current != null) {
            linkedList.insertAtHead(current.data);

            current = current.nextNode;
        }

        this.head = linkedList.head;
    }

    findKthFromEnd(k) {
        let i = 0;
        let map = new Map();

        let current = this.head;

        while (current != null) {
            map.set(i, current);

            i++;
            current = current.nextNode;
        }

        // In other approach, if I'd already have the length of the list,
        // I wouldn't need a map,
        // and the solution would be simpler an even more efficient...
        const length = i;
        const kth = length - k;

        return map.get(kth);
    }

    // O(n)
    removeDuplicates() {
        let set = new Set();

        let current = this.head;

        while (current != null) {
            set.add(current.data);

            if (current.nextNode && set.has(current.nextNode.data)) {
                current.nextNode = current.nextNode.nextNode;
            }

            current = current.nextNode;
        }

        // If additional memory wouldn't be allowed,
        // I'd have to traverse the list from head per each element...
    }

    deleteFromMiddle(data) {
        if (this.head && this.head.data === data) {
            this.deleteFromHead();

            return;
        }

        let current = this.head;

        while (current != null) {
            if (current.nextNode && current.nextNode.data === data) {
                current.nextNode = current.nextNode.nextNode;
            }

            current = current.nextNode;
        }
    }

    isCircular() {
        let isCircular = false;

        let current = this.head;

        while (current != null) {
            if (current.nextNode == this.head) {
                isCircular = true;

                break;
            }

            current = current.nextNode;
        }

        return isCircular;
    }

    toString() {
        let result = [];
        let current = this.head;

        while (current != null) {
            result.push(current.data);

            current = current.nextNode;
        }

        return result.join(', ');
    }
}

module.exports = LinkedList;